### ISSUE TITLE
Detection accuracy of the _cardio_learning models ??

### ISSUE THREAD

**Issue by cyberinQ** [Edoar-do/HuBERT-ECG#4]
Dear Eduardo, 
Thank you for publishing a thorough work. I tried to test all three  _cardio_lerarning models on a few real life ECG samples, but the results seem to be misleading. I have verified this with several cardiologists on several ECGs. What could I be doing wrong to cause misleading interpretations from the models?

I have divided the 10 second ecg into 2x 5 second segments, and printed out only the top 10 conditions detected in each segment.

<img width="1051" alt="Image" src="https://github.com/user-attachments/assets/14618eea-7e79-4ccd-a12e-08d995b7b08f" />
<img width="944" alt="Image" src="https://github.com/user-attachments/assets/6287ca92-7e7a-4c09-8cdd-e561a36f3471" />
<img width="977" alt="Image" src="https://github.com/user-attachments/assets/525959ce-27a1-4637-afdc-1b31907e1060" />

Thank you,
Sunil

**Comment by Edoar-do** [comment]
Hello @cyberinQ (Sunil), thank you for your interest in our work.

Are you using our code for inference? With no code it is hard for me to search for possible errors in the pipeline.

Nonetheless, first things that come in mind are:
- have you preprocessed the ECGs?
- are you using less than 12 leads? HuBERT-ECG processes all the 12 leads concatenated as I, II, III, aVR, aVL, aVF, V1,... V6 into a 1D signal. Different input might introduce distribution shifts
- you have to find empirically the optimal classification threshold that maximes your metrics, as we expressed performance in terms of AUROC
- those NaNs seem to indicate that the index of the classification head that gave that prediction does not match one of the columns in `cardio-learning-columns.pkl`. Is the indexing correct? Please note that you have to skip the first 3 columns as they are [filename, age, sex], not pathologies. This was made for consistencies with all the csv file used.  Hope the condition you're trying to predict actually is between Cardio-Learning's labels, otherwise the model will never predict that
- have you tried not to split your ECGs but to use TTA with `max` as aggregation? It gives better results

There are also a few "warnings" to keep in mind: given the training label distribution and the weak nature of the ground-truth (diagnoses might have been assigned erroneously (e.g. see PTB-XL paper on how they assign labels) or be incomplete), the model is likely to produce false positives as it is optimised for sensitivity and specificity (that's why literature proposes metrics like AUROC and rarely something based on precision), thus physicians are still fundamental to rule out mispredictions. Also, a further fine-tuning to better match the deployment distribution shouldn't be excluded a priori, as we intended HuBERT-ECG CardioLearning as a finetuned model with already some condition-related knowledge (in contrast to the more general self-supervised HuBERT-ECG) and not "the absolute" diagnoser to be used as-is

Hope I was helpful but please let me know if you need anything else.

Happy coding (Sorry for multiple edits)

**Comment by cyberinQ** [comment]
Dear Edoardo,

In the light of our discussions about false positives, and brainstorm with two senior cardiologists on my panel, it occurred to me that we could try delineating the problem into two parts:

Step A: Pure Signal Anomaly Detection (Engineering Perspective) - The model should analyze the raw ECG signals for any deviations in morphology, amplitude, and time jitter across all 12 leads. A signal-processing-driven approach, treating the ECG as a set of time-series waveforms, independent of medical knowledge. The goal is to flag waveform anomalies such as: Irregular PQRST complexes, Sudden shifts in amplitude, Phase distortions, Beat-to-beat variability outside normal thresholds, Noise, artifacts, and missing data, etc.

Step B: Clinical Interpretation (Medical Perspective) - Mapping the combinations of detected anomalies to clinical conditions mimicking a cardiologist’s diagnostic process, considering the interplay of waveform anomalies to infer potential diseases.
Example mappings: ST elevation in multiple leads → Acute MI,  Prolonged PR interval → First-degree AV block,  Wide QRS complex + tachycardia → Ventricular Tachycardia, etc..

Step A will ensure objective, physics-based anomaly detection without bias from pre-defined disease labels; and could independently be very useful for a clinician's productivity. Step B introduces clinical knowledge for meaningful interpretation as a second opinion. 

Assuming you agree to the above logic, how do we determine if the self-supervised pre-trained HuBert interally establishes some sort of "golden ECG" reference for itself, and whether it explicitly detects deviations in morphology, amplitude, and time jitter? 

I'm imagining, for example, it might be clustering ECG segments into learned representations, meaning it could be detecting out-of-distribution signals based on how far they deviate from its learned embeddings. 

Could we extract the learned embeddings of ECG signals and analyze how far a new 'test waveform' that we feed to analyze, deviates from those embeddings?

Thoughts?

Sunil
